### PR TITLE
Added Locals and Parameters to the eDSL

### LINKED ISSUE #706: Support locals and parameters in the Sequencing EDSL
The [Seq JSON spec](https://github.com/NASA-AMMOS/seq-json-schema/) supports passing sequence [parameters](https://github.com/NASA-AMMOS/seq-json-schema/blob/develop/schema.json#L651), and declaring [local variables](https://github.com/NASA-AMMOS/seq-json-schema/blob/develop/schema.json#L640). We need a way to allow declaring both parameters and locals in the sequencing EDSL, use them in the sequence, and properly output them in `.seq.json` as argument [symbols](https://github.com/NASA-AMMOS/seq-json-schema/blob/develop/schema.json#L392).

For example here is a proposed idea for how to do this in the EDSL. Psyche provided us one of their [sequence generation scripts](https://github.jpl.nasa.gov/PSYCHE/psyche-reusable-sequences/blob/develop/ep_swirl_preload.py), and [.seq.json output](https://github.jpl.nasa.gov/PSYCHE/psyche-reusable-sequences/blob/develop/ep/ep_swirl_preload/rs.ep_swirl_preload.r18.seq.json) we can draw inspiration from:

```ts
export default () =>
  Sequence.new({
    seqId: '',
    locals: [{ name: 'duration', type: 'FLOAT' }],
    parameters: [{ name: 'timeout', type: 'UINT' }],
    metadata: {},
    steps: [
      // Notice we use magic globals 'locals' and 'parameters' that bind to
      // the declared values in 'locals' and 'parameters' above.
      // This is just one idea.
      C.FSW_CMD_0({ arg0: locals.duration, arg1: parameters.timeout  })
    ],
  });
```

And the corresponding `.seq.json`:

```json
{
  "id": "",
  "metadata": {},
  "locals": [
    {
      "name": "duration",
      "type": "FLOAT"
    }
  ],
  "parameters": [
    {
      "name": "timeout",
      "type": "UINT"
    }
  ],
  "steps": [
    {
      "args": [
        { "name": "arg0", "type": "symbol", "value": "duration" },
        { "name": "arg1", "type": "symbol", "value": "timeout" }
      ],
      "stem": "FSW_CMD_0",
      "time": { "type": "COMMAND_COMPLETE" },
      "type": "command"
    }
  ]
}
```

### PR DESCRIPTION
**Pull request by goetzrrGit**
* **Tickets addressed:** Closes https://github.com/NASA-AMMOS/aerie/issues/706
* **Review:** By commit  
* **Merge strategy:** Merge (no squash)  

## Description

I've noticed that implementing `Locals` and `Parameters` in Monaco using a custom worker can be tricky. Currently, it's not possible to define a variable inside a JavaScript object and use it elsewhere in the same object. Additionally, valid TypeScript is required for the TS runner to execute the user's code, so using a placeholder in the sequence and ignoring the error during SeqJson generation is not an option.

Below is not possible in TS
```ts
export default () =>
  Sequence.new({
    seqId: '',
    locals: [{ name: 'duration', type: 'FLOAT' }],
    parameters: [{ name: 'timeout', type: 'UINT' }],
    metadata: {},
    steps: [
      // Notice we use magic globals 'locals' and 'parameters' that bind to
      // the declared values in 'locals' and 'parameters' above.
      // This is just one idea.
      C.FSW_CMD_0({ arg0: locals.duration, arg1: parameters.timeout  })
    ],
  });
```

To work around these restrictions, I've implemented using `const` outside of the sequence object and created helper functions to make defining parameters and locals more manageable. These can be used like variables in the command arguments, and the eDSL will try to link everything together as best as possible. However, going from seqJSON to eDSL can be tricky, as all the data is not in the same place. To help with this, I've implemented more extensive error checking for going to and from seqJSON.

```ts
const LOCALS = BUILD_LOCALS(
    VARIABLE('temp','FLOAT')
  )
  
  const PARAMETERS = BUILD_PARAMETERS(
    VARIABLE('sugar','INT')
  )
  
  export default () =>
    Sequence.new({
      seqId: '',
      metadata: {},
      locals: MAP_VARIABLES(LOCALS),
      parameters: MAP_VARIABLES(PARAMETERS),
      steps: [
        C.PREHEAT_OVEN({
          temperature: LOCALS.temp
        }),
        C.PREPARE_LOAF({
          tb_sugar: PARAMETERS.sugar,
          gluten_free: 'FALSE',
        }),
      ],
    });
```

I've included examples of errors that may occur when converting between eDSL and seqJSON, which should make it easier to identify and fix any issues. Do you want to keep this version of the description?

### **eDSL to SeqJSON error checking**

```ts
const LOCALS = BUILD_LOCALS(
  VARIABLE('temperature','STRING').ENUM_NAME('TEMPERATURE_ARRAY'),
)

const PARAMETERS = BUILD_PARAMETERS(
  VARIABLE('arm_position','ENUM').ALLOWABLE_RANGES([{min: 1, max: 2}]),
  VARIABLE('arm_rotation','STRING').ALLOWABLE_RANGES([{min: 1, max: 2}])
)

export default () =>
  Sequence.new({
    seqId: '',
    metadata: {},
    locals: MAP_VARIABLES(LOCALS),
    parameters: MAP_VARIABLES(PARAMETERS),
    immediate_commands : [PREHEAT_OVEN({temperature : LOCALS.temptemperature})]
  });
```

Errors caught and displayed in SeqJSON
```json
{
  "id": "",
  "immediate_commands": [
    {
      "args": [
        {
          "name": "$$ERROR$$",
          "type": "undefined",
          "value": "temperature is an unknown value"
        }
      ],
      "stem": "PREHEAT_OVEN"
    }
  ],
  "locals": [
    {
      "enum_name": "TEMPERATURE_ARRAY",
      "name": "$$ERROR$$: 'enum_name: TEMPERATURE_ARRAY' is not required for non-ENUM type.",
      "type": "STRING"
    }
  ],
  "metadata": {},
  "parameters": [
    {
      "allowable_ranges": [
        {
          "max": 2,
          "min": 1
        }
      ],
      "name": "$$ERROR$$: 'enum_name' is required for ENUM type.",
      "type": "ENUM"
    },
    {
      "allowable_ranges": [
        {
          "max": 2,
          "min": 1
        }
      ],
      "name": "$$ERROR$$: 'allowable_ranges' is not required for STRING type.",
      "type": "STRING"
    }
  ]
}
```

### **From SeqJSON to eDSL Error checking**

```json
{
  "id": "",
  "metadata": {},
  "parameters": [
    {
      "name": "sugar",
      "type": "INT"
    }
  ],
  "steps": [
    {
      "args": [
        {
          "name": "tb_sugar",
          "type": "symbol",
          "value": "sugarrrrr"
        },
        {
          "name": "gluten_free",
          "type": "string",
          "value": "FALSE"
        }
      ],
      "stem": "PREPARE_LOAF",
      "time": {
        "type": "COMMAND_COMPLETE"
      },
      "type": "command"
    }
  ]
}
```

Errors caught and displayed in the eDSL
```ts
const PARAMETERS = BUILD_PARAMETERS(
  VARIABLE('sugar','INT')
)

export default () =>
  Sequence.new({
    seqId: '',
    metadata: {},
    parameters: MAP_VARIABLES(PARAMETERS),
    steps: [
      C.PREPARE_LOAF({
        tb_sugar: UNKNOWN.sugarrrrr, //ERROR:  variable 'sugarrrrr' is not defined as a local or parameter"
        gluten_free: 'FALSE',
      }),
    ],
  });
```

## Verification
Added new e2e test to check all the errors for `locals` and `parameters`. All pass

## Documentation
This is probably a good write up to include in the wiki

## Future work
Globals are next

### CODE REVIEW

File: sequencing-server/src/lib/codegen/CommandTypeCodegen.ts [hunk 1]
```diff
@@ -57,7 +57,7 @@ export const Hardwares = {\n${dictionary.hwCommands
     .map(hwCommands => `\t\t${hwCommands.stem}: ${hwCommands.stem},\n`)
     .join('')}};
 
-Object.assign(globalThis, { A:A, R:R, E:E, C:Object.assign(Commands, STEPS), Sequence}, Hardwares, Immediates);
+Object.assign(globalThis, { A:A, R:R, E:E, C:Object.assign(Commands, STEPS), Sequence, VARIABLE, BUILD_LOCALS, BUILD_PARAMETERS, MAP_VARIABLES}, Hardwares, Immediates);
```
Comment (cohansen) [inline sequencing-server/src/lib/codegen/CommandTypeCodegen.ts:60]: How do we feel about prefacing our functions with an identifier to distinguish them from our other objects?
Something like `functions.BUILD_PARAMETER(...)`?

File: sequencing-server/src/lib/codegen/CommandEDSLPreface.ts [hunk 2]
```diff
@@ -7,6 +7,15 @@ export enum TimingTypes {
...
 
+export type VariableOptions = {
+  name: string;
+  type: 'FLOAT' | 'INT' | 'STRING' | 'UINT' | 'ENUM';
+  enum_name?: string | undefined;
+  allowable_values?: unknown[] | undefined;
+  // @ts-ignore : 'VariableRange' found in JSON Spec
+  allowable_ranges?: VariableRange[] | undefined;
+};
+
```
Comment (cohansen) [inline sequencing-server/src/lib/codegen/CommandEDSLPreface.ts:20]: I think you could add an extra type here that would require the `enum_name` field if the type was set to `ENUM`, something like:
```
type DefaultVariableOptions = {
  name: string;
  type: 'FLOAT' | 'INT' | 'STRING' | 'UINT';
  allowable_values?: unknown[] | undefined;
  allowable_ranges?: number[] | undefined;
}

type EnumVariableOptions = {
  name: string;
  type: 'ENUM'
  enum_name: string;
};

type VariableOptions = DefaultVariableOptions | EnumVariableOptions;

const testVar: VariableOptions = { // Type '{ name: string; type: "ENUM"; }' is not assignable to type 'VariableOptions'. Property 'enum_name' is missing in type '{ name: string; type: "ENUM"; }' but required in type 'EnumVariableOptions'.
  name: 'test',
  type: 'ENUM'
};
```

File: sequencing-server/src/lib/codegen/CommandEDSLPreface.ts [hunk 3]
```diff
@@ -238,6 +247,19 @@ declare global {
   type F32 = F<32>;
   type F64 = F<64>;
 
+  function VARIABLE(name: string, type: 'FLOAT' | 'INT' | 'STRING' | 'UINT' | 'ENUM'): Variable;
```
Comment (cohansen) [inline sequencing-server/src/lib/codegen/CommandEDSLPreface.ts]: Should this list of valid types be a const somewhere since it's reused? At least it would be easier to modify later.

**Review by cohansen — CHANGES_REQUESTED** [review]
Good job Ryan, it looks good and was easy to use in the UI once I knew the conventions. I think adding some comments that propagate into the editor would be really useful.

File: sequencing-server/src/lib/codegen/CommandEDSLPreface.ts (hunk 2 above)
Comment (goetzrrGit) [reply sequencing-server/src/lib/codegen/CommandEDSLPreface.ts:20]: It starts to get tricky as I am not sure how to limit the available functions. For example, If I provide a`DefaultVariableOptions ` I still have access to the `ENUM_NAME` function to set the enum name thus getting us to the error of defining an enum name without the enum type. Ideally, I would not allow the user to have access to that function.

`VARIABLE('duration', 'FLOAT').ENUM_NAME("DURATION_TYPE")`

File: sequencing-server/src/lib/codegen/CommandEDSLPreface.ts (hunk 3 above)
Comment (goetzrrGit) [reply sequencing-server/src/lib/codegen/CommandEDSLPreface.ts]: Nice idea, I just pulled it out of the interface and make it const.

File: sequencing-server/src/lib/codegen/CommandTypeCodegen.ts (hunk 1 above)
Comment (goetzrrGit) [reply sequencing-server/src/lib/codegen/CommandTypeCodegen.ts:60]: Are you think the user will type `functions` and get a list of available eDSL functions in the content assist?

Comment (cohansen) [reply sequencing-server/src/lib/codegen/CommandTypeCodegen.ts:60]: Yeah that was the idea I had, I'm not sure if it would make the assist any better but maybe it would help prompt the user more. Just a random thought!

File: sequencing-server/src/lib/codegen/CommandEDSLPreface.ts (hunk 2 above)
Comment (cohansen) [reply sequencing-server/src/lib/codegen/CommandEDSLPreface.ts:20]: Ah, yeah that makes sense. I also wrote this comment before I scrolled down and saw the `Variable` class.

**Comment by dyst5422** [comment]
@goetzrrGit Whats your predicted timeline on merging this? I want to play around with alternative API's like what we had in the issue discussion as an exercise in "what other options do we have".

Specifically, I'd like to experiment a little with what we can do in terms of making the inputs look more like inputs a la https://github.com/NASA-AMMOS/aerie/issues/706#issuecomment-1452263005.

If you could delay merging till Monday, I think I could look into it before then.

I'm thinking something like
```
export default () =>
  Sequence.new({
    seqId: '',
    locals: [{ name: 'duration', type: 'FLOAT' }],
    parameters: [{ name: 'timeout', type: 'UINT' }],
    metadata: {},
    steps: ({ locals, parameters}) => ([
      C.FSW_CMD_0({ arg0: locals.duration, arg1: parameters.timeout  })
    ]),
  });
```

where the constructor would pull the locals and parameters off and pass them to the steps function to build out the resulting seqjson. Not sure if it would work yet, but I think it might be an interesting API to compare with your solution for the users.

File: sequencing-server/src/lib/codegen/CommandTypeCodegen.ts (hunk 1 above)
Comment (goetzrrGit) [reply sequencing-server/src/lib/codegen/CommandTypeCodegen.ts:60]: I honestly was thinking about something similar recently and was more on the fence if users would want to have to type more, which seems like we are trying to limit that as much as possible. It might be a good idea to have a discussion about it with the team cause it will help out new users.

I was also thinking about being more explicit so you exactly what you are typing

```ts
eDSL.step.<c/a/e>.<command>
eDSL.step.ground_block
eDSL.step.ground_event

eDSL.hdw.<hardware_command>

eDSL.immediate.<commands>

eDSL.functions.build_locals
eDSL.functions.map_variables
eDSL.functions.time.<time helper functions>
```

Comment (goetzrrGit) [reply sequencing-server/src/lib/codegen/CommandTypeCodegen.ts:60]: maybe typing `eDSL` is too repetitive.

File: sequencing-server/src/lib/codegen/CommandEDSLPreface.ts (hunk 2 above)
Comment (goetzrrGit) [reply sequencing-server/src/lib/codegen/CommandEDSLPreface.ts:20]: No worries, It was a good idea.

**Comment by goetzrrGit** [comment]
@dyst5422 Hi Dylan, I hope you are doing well. I hope to get this merge in today just making some small changes and making sure the test all pass. 

@camargo do you want to delay until Monday as I know you want to release on that day

**Comment by cohansen** [comment]
> @dyst5422 Hi Dylan, I hope you are doing well. I hope to get this merge in today just making some small changes and making sure the test all pass.
> 
> @camargo do you want to delay until Monday as I know you want to release on that day

Chris mentioned that he was off on release days. So the next release is not next Tuesday but the following!

**Comment by dyst5422** [comment]
@goetzrrGit I think I have the basic structure of the necessary parts to make this possible in this [Playground Link](https://www.typescriptlang.org/play?#code/FAFwngDgpgBAagQwE4EsECMA2UAiUDGmyCIKA9gHYwC8MA3jBQgLZQBcMAziKhQOYAaGOGgduvPjAC+AbmDAA9ApgAVABYpOMTTCidOUCqQSZMYGGQCuSGAGUoAR0uH8UAHQUoAd2AAzSxT4pJQw+JTiCCgUADzAMDAKAFSqarBhFNwwJMKpXCDIIDCpSLAgUKZaKrbCZDAA1lBQEDmwTEhIZF56hSKwEGT6KFhQcfHJBrBqICAQnGxKXotuvZz4qBAgRPxuZEh8CgAmZPicCmoIFAfoZGR1CiXYCAYAtBRkZacraygbzwCszwADG4psxMABidLcZ69Z4QYisMpITijGDJdQ6CAlABuhhAWhAuQAqgBJGC+DrMGAHTQQIhgKKSThkREafgwTAoBr0ADaDTAYh4jIAuoKJFJtBkyggDhZfPR6lAwABGDjYkzOZVCfkAJjVGucOukqMSClGUMKABljiYtFAAB5lS5aABKUBllDMAEF2ggwNFEKgMNg8IRiMEKAA+GgwHnCoTm8KFAAKCKgSLtjsMB1d7qOFG9vv9gbQw1DRCQJHIUZjcYEwEjAAoyBs5vRRpgbZhOAB+DjW-C2+vxeGVxFQZF9mCpsfpiecYd5Jpt8SMuMwAA+MEbzdbHDoqISiQAYhd8OYvH6ajBmAhmlYbMhK+YyPKyOgAFYEfHXi4WT-fjAXgoISjAsLATyKuYFyyr0WRaOqmDOIeaI5DonhQDm17oLAUScp4spRNweZyrojqVkEmG-kU5TQDYcEmGgWjMjk2TMJQCBhEBlAAOSFBoBylLkASCUgZiMsIkCwDhbKwbkjFPChpqHp2g7dvusYkpcDqSjAA62jyFCWMwOFIMK8EwFpgn2jyPFMKwPHCqKlnaTZfFSY5xrxCOaYZhpPJWTpUTTr586GcZpnmZBgVufZUCOc5MW2b0nlSKMUgAJQ0NGq78HGGXAJl7bxEkyTHrs+ASYSaS7CUQSSdA5K7C0MAiVAvhRFRg4GGiZrxBaHJdlotAtvibiqbaMA9j2sZRVoAR1G8XhUJB+ndnI-VJjAo7gRmMajZwbg7eOyJTTN66QQtS0rVoM67fOG2hFtE3dgAsneMYvYdJQHJYrg7px+BCIJYaVhGWXUNGB7eUewhaCgfBvCUNHIzhg6WD1XQ8cjuJIOYzB-WoMAAHIAPIqNSZASV45yFFjyNlKYElVEBEElEI6CWIUIE3ggDRaL4zV8DcspEgAGkprPY607xZEElgmGY21PJwVW5HBWIthO4CkdV1IEBWVaUBzXMtTTZDYProNGxQksYnaGTWHorMwHw6ZZDAJQysGpRSbpevEbM5IBEE1ZKcoh5KDAAAC+LPAjSMjDDgM8iDhsRh44HmbQafhtWj3xCUIDWCt+D4I9UhCHQEqQQwAWubpa2cOFJkTnNLnWTyABEcVd05HBJV3vR9zI0gZY9A3HXOyLvc0tBTxmbg-X9UAA2XwMG3nlAQ1DkfKD+Ce7JME7SQQCAY7A9OwLj+OEyT5OU9TtNSwz5T4ZILNdFk7MwJz3OFLefmTUbDCzIKLCWMMo6x04PHRGR9Dwp1zmDasmdWDZytunfOh4i4lzluXdKVca5aDrjFRuQ0W6RQsoPXu-cO4Om7sPYUo9MoTy2oHYaDUoCvgsK2Nw7CaDUFoDxfwgQIw8Smjwsa7DGwMC+hwL6s8hAL3nBwZRM8PpFQ4AdPhZRZiPRwUgKgsihpaN4V9IQ-DZCFXkCI0OIQiQAGYdTRAGsTMiTosK5T4E2OKHBiYZQ0r4kmQhegcB4pYJx4iJTQ09umXBDA4ohKkmEiJOoolyDSn4EOEYYDHgAGwABYXFbTcQ6DxzEhT8B8eBPxASFRBOJkk0QMBhGFKicVWJxdDH1PAk09gLTfBtOkBk+QA0PiFFoBaSsUQZEdhMbGRxOpGw8V+sg3iGUhD5IKcs0grArAgB4hleMox2EcBkYNNSC5lazj2kVSG24eQqSGm4VZNtFzxC+ssFAeyuajGFAVFhozwgW3cJ2PgjZxnjyAA)

Note: This requires a feature that landed in TS 5.0 (bleeding edge) so we would need to update our TS version for this.

I had to do a minimal example because I didn't have time to figure out the entirety of getting a properly built command typescript out to mess with. I think all the parts to put it into the context of our sequence typescript are there though.

Note: We're doing a bit of tricky stuff here where we're telling the compiler one thing, and doing another. Specifically, we are telling it we're passing to the step function a map of {[name: string]: string} which indicates to the users the mapping from variable name to type, however, at runtime we're ACTUALLY passing the full variable declaration object.


TLDR: this will allow us to do exactly the syntax I had above
```ts
export default () =>
  Sequence.new({
    seqId: '',
    locals: [{ name: 'duration', type: 'FLOAT' }],
    parameters: [{ name: 'timeout', type: 'UINT' }],
    metadata: {},
    steps: ({ locals, parameters}) => ([
      C.FSW_CMD_0({ arg0: locals.duration, arg1: parameters.timeout  })
    ]),
  });
```

Of course, we want the functional version of steps to be optional, so we'll have to handle the appropriate union type `Step[] | (...) => Step[]` and runtime check `typeof steps === 'function'`, but we have prior art already for that

**Comment by goetzrrGit** [comment]
@dyst5422 Wow, I kneel before the TS King!!! I am not worthy...
Let me play around and see if I can understand the tricky you came up with.

**Comment by dyst5422** [comment]
> @dyst5422 Wow, I kneel before the TS King!!! I am not worthy...
> Let me play around and see if I can understand the tricky you came up with.

@goetzrrGit Reload the playground from the link, because I just realized I had to do a quick small update

And let me know if you need some help understanding it - we're definitely at the edge of what the TS type system can do, so it's most certainly not straightforward.

**Comment by dyst5422** [comment]
Additional thoughts: we can make the locals and parameters a kv map from names to types of we'd like as well

`locals: { duration: Float, timeout: U32 }`

If we make our primitive types constants/enums and construct the seqjson form in the constructor if we'd like as well (would simplify the type magic going on in that playground example as well).

**Comment by goetzrrGit** [comment]
Going to close in favor of TS 5 https://github.com/NASA-AMMOS/aerie/pull/890
